Add metrics spec supplementary guidelines

reyang opened the pull request:
Related to #1873 and #1891.

If you find it hard to look at the raw markdown file, here is a rendered version https://github.com/reyang/opentelemetry-specification/blob/reyang/guideline/specification/metrics/supplementary-guidelines.md.

utpilla commented on specification/metrics/supplementary-guidelines.md:
```diff
@@ -0,0 +1,182 @@
+  * ProcessId = `1001`, ThreadId = `2`, PageFaults = `53`
+  * ProcessId = `1001`, ThreadId = `3`, PageFaults = `5`
+
+If we export the metrics using **Cumulative Temporality**, it is quite
```
Should this be Delta Temporality and the next paragraph be Cumulative?

reyang replied:
Nope, this is Cumulative (we're doing Cumulative->Cumulative here).

The next paragraph is about Cumulative->Delta conversion.

utpilla commented on specification/metrics/supplementary-guidelines.md:
```diff
@@ -0,0 +1,182 @@
+* If the maximum value is 20 during (T<sub>0</sub>, T<sub>2</sub>] and the
+  maximum value is also 20 during (T<sub>0</sub>, T<sub>3</sub>], we wouldn't
+  know what the maximum value is during (T<sub>2</sub>, T<sub>3</sub>], unless
+  we know that there is no value (count = 0).
```
Maybe reword this to be more specific?

"unless we know that there is no value during (T<sub>2</sub>, T<sub>3</sub>] (count = 0) in which case the maximum value during (T<sub>2</sub>, T<sub>3</sub>] will be 0."

reyang replied:
Probably no? I think if count=0, the maximum value technically should not be exist, or it can be NaN.

utpilla replied:
If the maximum value shouldn't technically exist, then maybe we should get rid of the "unless we know ..." part.

utpilla commented on specification/metrics/supplementary-guidelines.md:
```diff
@@ -0,0 +1,182 @@
+
+If we choose **Cumulative Temporality**, the SDK **has to track what has
+happened prior to the latest collection/export cycle**, in this worst case, the
+SDK **will remember what has happened since the ever beginning of the process**.
```
nit: SDK **will have to remember what has happened ever since the beginning of the process**.

reyang replied:
Fixed, thanks!

jsuereth reviewed (commented):
This generally looks very good.  I need to take a second pass on grammar/phrasing (That's my own worst enemy), but the content looks like a great addition/treatment of the issue for implementers.

One thing I'd like to call out though is a suggestion around Memory Management and Allocations.  While it's impossible to remove all allocations on the hot path of synchronous metric collection, particularly if using Baggage, SDK authors should be encouraged to reduce memory contention *in the hot path*, and pay most overhead at startup.   This aligns with OTel's vision of "least disruption to runtime", which shows up in recommendations around error-reporting during operation and likely should also apply to things like memory usage.

I like the idea of limiting memory as suggested here, just want to also reinforce this "allocate ahead of time and reuse" notion, which we actually expose a bit more directly in Java (and I need to annotate).


Specifically, one scenario to call out (more important for Logs/Traces than metrics) is how a system behaves when memory can no longer be allocated.   Do we still get observability or is it gone?

reyang commented:
> I like the idea of limiting memory as suggested here, just want to also reinforce this "allocate ahead of time and reuse" notion, which we actually expose a bit more directly in Java (and I need to annotate).

I've updated the PR to cover the pre-allocation for garbage collected languages. PTAL.
 
> Specifically, one scenario to call out (more important for Logs/Traces than metrics) is how a system behaves when memory can no longer be allocated. Do we still get observability or is it gone?

I've updated the PR trying to scratch the surface. If we need to dig deeper, a separate PR might be better?

alanwest commented on specification/metrics/supplementary-guidelines.md:
```diff
@@ -0,0 +1,251 @@
+You can see that we are performing Delta->Cumulative conversion, and the SDK
+**has to track what has happened prior to the latest collection/export cycle**,
+in the worst case, the SDK **will have to remember what has happened since the
+ever beginning of the process**.
```
```suggestion
beginning of the process**.
```

cijothomas commented on specification/metrics/supplementary-guidelines.md:
```diff
@@ -0,0 +1,251 @@
+If we export the metrics using **Delta Temporality**:
+
+* (T<sub>0</sub>, T<sub>1</sub>]
+  * dimensions: {verb = `GET`, status = `200`}, count: `2`, min: `50 (ms)`, max:
```
nit: Histogram also has Sum

reyang replied:
Yeah, I think I omitted many things here intentionally so the reader could focus on the key points.

jmacd reviewed (approved):
❤️ ❤️ ❤️
